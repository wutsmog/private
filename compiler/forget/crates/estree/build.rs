use estree_codegen::estree;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../estree-codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../estree-codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../estree-codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../estree-codegen");

    let src = estree();
    std::fs::write("src/generated.rs", src).unwrap();
}
