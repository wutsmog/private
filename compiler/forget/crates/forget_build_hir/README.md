# Build-HIR

This crate converts from `forget_estree` into Forget's HIR format as the first phase of compilation.