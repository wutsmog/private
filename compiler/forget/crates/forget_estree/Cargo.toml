[package]
name = "forget_estree"
version = "0.1.0"
publish               = false
authors.workspace     = true
description.workspace = true
edition.workspace     = true
homepage.workspace    = true
keywords.workspace    = true
license.workspace     = true
repository.workspace  = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bumpalo = { workspace = true }
insta = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
static_assertions = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }