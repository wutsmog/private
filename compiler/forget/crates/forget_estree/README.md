# forget_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

This crate is intended as the main interchange format with outside code. A typical integration with Forget
will look as follows:

1. Host Compiler parses into the host AST format.
2. Host Compiler converts into `forget_estree`.
3. Host Compiler invokes Forget to compile the input, which (conceptually) 
   returns the resulting code in `forget_estree` format.
4. Host Compiler convert back from `forget_estree` to its host AST format.

Because Forget is intended to support JavaScript-based toolchains, `forget_estree` is designed to support
accurate serialization to/from estree-compatible JSON. We may also support the Babel AST format 
(a variant of ESTree) as well, depending on demand.