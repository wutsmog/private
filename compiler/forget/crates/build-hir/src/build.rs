use bumpalo::collections::{String, Vec};
use estree::{
    AssignmentTarget, BinaryExpression, BlockStatement, Expression, ForInit, ForStatement,
    FunctionDeclaration, IfStatement, JsValue, Literal, Pattern, Statement,
    VariableDeclarationKind,
};
use hir::{
    ArrayElement, BlockKind, BranchTerminal, Environment, ForTerminal, Function, GotoKind,
    IdentifierOperand, InstrIx, InstructionKind, InstructionValue, LValue, LoadGlobal, LoadLocal,
    Operand, PrimitiveValue, TerminalValue,
};

use crate::{
    builder::{Binding, Builder, LoopScope},
    error::DiagnosticError,
    BuildDiagnostic, ErrorSeverity,
};

/// Converts a React function in ESTree format into HIR. Returns the HIR
/// if it was constructed sucessfully, otherwise a list of diagnostics
/// if the input could be not be converted to HIR.
///
/// Failures generally include nonsensical input (`delete 1`) or syntax
/// that is not yet supported.
pub fn build<'a>(
    env: &'a Environment<'a>,
    fun: FunctionDeclaration,
) -> Result<&'a mut Function<'a>, BuildDiagnostic> {
    let mut builder = Builder::new(env);

    match fun.function.body {
        Some(estree::FunctionBody::BlockStatement(body)) => {
            lower_block_statement(env, &mut builder, *body)?
        }
        Some(estree::FunctionBody::Expression(body)) => {
            lower_expression(env, &mut builder, body)?;
        }
        None => {
            return Err(BuildDiagnostic::new(
                DiagnosticError::EmptyFunction,
                ErrorSeverity::InvalidSyntax,
                fun.range,
            ));
        }
    }

    let mut params = Vec::with_capacity_in(fun.function.params.len(), &env.allocator);
    for param in fun.function.params {
        match param {
            Pattern::Identifier(param) => {
                let identifier = lower_identifier_for_assignment(
                    env,
                    &mut builder,
                    InstructionKind::Let,
                    *param,
                )?;
                params.push(identifier);
            }
        }
    }

    // In case the function did not explicitly return, terminate the final
    // block with an explicit `return undefined`. If the function *did* return,
    // this will be unreachable and get pruned later.
    let implicit_return_value = builder.push(InstructionValue::Primitive(hir::Primitive {
        value: PrimitiveValue::Undefined,
    }));
    builder.terminate(
        TerminalValue::Return(hir::ReturnTerminal {
            value: Operand {
                ix: implicit_return_value,
                effect: None,
            },
        }),
        hir::BlockKind::Block,
    );

    let body = builder.build()?;
    Ok(env.alloc(Function {
        id: fun
            .function
            .id
            .map(|id| String::from_str_in(&id.name, &env.allocator)),
        body,
        params,
        is_async: fun.function.is_async,
        is_generator: fun.function.is_generator,
    }))
}

fn lower_block_statement<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    stmt: BlockStatement,
) -> Result<(), BuildDiagnostic> {
    for stmt in stmt.body {
        lower_statement(env, builder, stmt, None)?;
    }
    Ok(())
}

/// Convert a statement to HIR. This will often result in multiple instructions and blocks
/// being created as statements often describe control flow.
fn lower_statement<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    stmt: Statement,
    label: Option<String<'a>>,
) -> Result<(), BuildDiagnostic> {
    match stmt {
        Statement::BlockStatement(stmt) => {
            lower_block_statement(env, builder, *stmt)?;
        }
        Statement::BreakStatement(stmt) => {
            let block = builder.resolve_break(stmt.label.as_ref())?;
            builder.terminate(
                TerminalValue::Goto(hir::GotoTerminal {
                    block,
                    kind: GotoKind::Break,
                }),
                BlockKind::Block,
            );
        }
        Statement::ContinueStatement(stmt) => {
            let block = builder.resolve_continue(stmt.label.as_ref())?;
            builder.terminate(
                TerminalValue::Goto(hir::GotoTerminal {
                    block,
                    kind: GotoKind::Continue,
                }),
                BlockKind::Block,
            );
        }
        Statement::ReturnStatement(stmt) => {
            let ix = match stmt.argument {
                Some(argument) => lower_expression(env, builder, argument)?,
                None => builder.push(InstructionValue::Primitive(hir::Primitive {
                    value: PrimitiveValue::Undefined,
                })),
            };
            builder.terminate(
                TerminalValue::Return(hir::ReturnTerminal {
                    value: Operand { ix, effect: None },
                }),
                BlockKind::Block,
            );
        }
        Statement::ExpressionStatement(stmt) => {
            lower_expression(env, builder, stmt.expression)?;
        }
        Statement::EmptyStatement(_) => {
            // no-op
        }
        Statement::VariableDeclaration(stmt) => {
            let kind = match stmt.kind {
                VariableDeclarationKind::Const => InstructionKind::Const,
                VariableDeclarationKind::Let => InstructionKind::Let,
                VariableDeclarationKind::Var => {
                    return Err(BuildDiagnostic::new(
                        DiagnosticError::VariableDeclarationKindIsVar,
                        ErrorSeverity::Unsupported,
                        stmt.range,
                    ));
                }
            };
            for declaration in stmt.declarations {
                if let Some(init) = declaration.init {
                    let value = lower_expression(env, builder, init)?;
                    lower_assignment(
                        env,
                        builder,
                        kind,
                        AssignmentTarget::Pattern(declaration.id.into()),
                        value,
                    )?;
                } else {
                    if let Pattern::Identifier(id) = declaration.id {
                        // TODO: handle unbound variables
                        let binding = builder.resolve_binding(&id)?;
                        let identifier = match binding {
                            Binding::Local(identifier) => identifier,
                            _ => {
                                return Err(BuildDiagnostic::new(
                                    DiagnosticError::VariableDeclarationBindingIsNonLocal,
                                    ErrorSeverity::Invariant,
                                    id.range,
                                ));
                            }
                        };
                        builder.push(InstructionValue::DeclareLocal(hir::DeclareLocal {
                            lvalue: LValue {
                                identifier: IdentifierOperand {
                                    identifier,
                                    effect: None,
                                },
                                kind,
                            },
                        }));
                    }
                }
            }
        }
        Statement::IfStatement(stmt) => {
            // block for what follows the if statement, though this may
            // not be reachable
            let fallthrough_block = builder.reserve(BlockKind::Block);

            let IfStatement {
                test,
                consequent,
                alternate,
                ..
            } = *stmt;

            let consequent_block = builder.enter(BlockKind::Block, |builder| {
                lower_statement(env, builder, consequent, None)?;
                Ok(TerminalValue::Goto(hir::GotoTerminal {
                    block: fallthrough_block.id,
                    kind: GotoKind::Break,
                }))
            })?;

            let alternate_block = builder.enter(BlockKind::Block, |builder| {
                if let Some(alternate) = alternate {
                    lower_statement(env, builder, alternate, None)?;
                }
                Ok(TerminalValue::Goto(hir::GotoTerminal {
                    block: fallthrough_block.id,
                    kind: GotoKind::Break,
                }))
            })?;

            let test = lower_expression(env, builder, test)?;
            let terminal = TerminalValue::If(hir::IfTerminal {
                test: Operand {
                    ix: test,
                    effect: None,
                },
                consequent: consequent_block,
                alternate: alternate_block,
                fallthrough: Some(fallthrough_block.id),
            });
            builder.terminate_with_fallthrough(terminal, fallthrough_block);
        }
        Statement::ForStatement(stmt) => {
            let ForStatement {
                init,
                test,
                update,
                body,
                ..
            } = *stmt;

            // Block for the loop's test condition
            let test_block = builder.reserve(BlockKind::Loop);

            // Block for code following the loop
            let fallthrough_block = builder.reserve(BlockKind::Block);

            let init_block = builder.enter(BlockKind::Loop, |builder| {
                if let Some(ForInit::VariableDeclaration(decl)) = init {
                    lower_statement(env, builder, Statement::VariableDeclaration(decl), None)?;
                    Ok(TerminalValue::Goto(hir::GotoTerminal {
                        block: test_block.id,
                        kind: GotoKind::Break,
                    }))
                } else {
                    Err(BuildDiagnostic::new(
                        DiagnosticError::ForStatementIsMissingInitializer,
                        ErrorSeverity::Todo,
                        None,
                    ))
                }
            })?;

            let update_block = update
                .map(|update| {
                    builder.enter(BlockKind::Loop, |builder| {
                        lower_expression(env, builder, update)?;
                        Ok(TerminalValue::Goto(hir::GotoTerminal {
                            block: test_block.id,
                            kind: GotoKind::Break,
                        }))
                    })
                })
                .transpose()?;

            let body_block = builder.enter(BlockKind::Block, |builder| {
                let loop_ = LoopScope {
                    label,
                    continue_block: update_block.unwrap_or(test_block.id),
                    break_block: fallthrough_block.id,
                };
                builder.enter_loop(loop_, |builder| {
                    lower_statement(env, builder, body, None)?;
                    Ok(TerminalValue::Goto(hir::GotoTerminal {
                        block: update_block.unwrap_or(test_block.id),
                        kind: GotoKind::Continue,
                    }))
                })
            })?;

            let terminal = TerminalValue::For(ForTerminal {
                body: body_block,
                init: init_block,
                test: test_block.id,
                fallthrough: fallthrough_block.id,
                update: update_block,
            });
            builder.terminate_with_fallthrough(terminal, test_block);

            if let Some(test) = test {
                let test_value = lower_expression(env, builder, test)?;
                let terminal = TerminalValue::Branch(BranchTerminal {
                    test: Operand {
                        ix: test_value,
                        effect: None,
                    },
                    consequent: body_block,
                    alternate: fallthrough_block.id,
                });
                builder.terminate_with_fallthrough(terminal, fallthrough_block);
            } else {
                return Err(BuildDiagnostic::new(
                    DiagnosticError::ForStatementIsMissingTest,
                    ErrorSeverity::Todo,
                    stmt.range,
                ));
            }
        }
        _ => todo!("Lower {stmt:#?}"),
    }
    Ok(())
}

/// Converts an ESTree Expression into an HIR InstructionValue. Note that while only a single
/// InstructionValue is returned, this function is recursive and may cause multiple instructions
/// to be emitted, possibly across multiple basic blocks (in the case of expressions with control
/// flow semenatics such as logical, conditional, and optional expressions).
fn lower_expression<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    expr: Expression,
) -> Result<InstrIx, BuildDiagnostic> {
    let value = match expr {
        Expression::Identifier(expr) => {
            // TODO: handle unbound variables
            let binding = builder.resolve_binding(&expr)?;
            match binding {
                Binding::Local(identifier) => {
                    let place = IdentifierOperand {
                        effect: None,
                        identifier,
                    };
                    InstructionValue::LoadLocal(LoadLocal { place })
                }
                Binding::Module(..) | Binding::Global => InstructionValue::LoadGlobal(LoadGlobal {
                    name: String::from_str_in(&expr.name, &env.allocator),
                }),
            }
        }
        Expression::Literal(expr) => InstructionValue::Primitive(hir::Primitive {
            value: lower_primitive(env, builder, *expr),
        }),
        Expression::ArrayExpression(expr) => {
            let mut elements = Vec::with_capacity_in(expr.elements.len(), &env.allocator);
            for expr in expr.elements {
                let element = match expr {
                    Some(estree::ExpressionOrSpread::SpreadElement(expr)) => {
                        Some(ArrayElement::Spread(Operand {
                            ix: lower_expression(env, builder, expr.argument)?,
                            effect: None,
                        }))
                    }
                    Some(estree::ExpressionOrSpread::Expression(expr)) => {
                        Some(ArrayElement::Place(Operand {
                            ix: lower_expression(env, builder, expr)?,
                            effect: None,
                        }))
                    }
                    None => None,
                };
                elements.push(element);
            }
            InstructionValue::Array(hir::Array { elements })
        }

        Expression::AssignmentExpression(expr) => match expr.operator {
            estree::AssignmentOperator::Equals => {
                let right = lower_expression(env, builder, expr.right)?;
                return Ok(lower_assignment(
                    env,
                    builder,
                    InstructionKind::Reassign,
                    expr.left,
                    right,
                )?);
            }
            _ => todo!("lower assignment expr {:#?}", expr),
        },

        Expression::BinaryExpression(expr) => {
            let BinaryExpression {
                left,
                operator,
                right,
                ..
            } = *expr;
            let left = lower_expression(env, builder, left)?;
            let right = lower_expression(env, builder, right)?;
            InstructionValue::Binary(hir::Binary {
                left: Operand {
                    ix: left,
                    effect: None,
                },
                operator,
                right: Operand {
                    ix: right,
                    effect: None,
                },
            })
        }

        _ => todo!("Lower expr {expr:#?}"),
    };
    Ok(builder.push(value))
}

fn lower_assignment<'a>(
    env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    kind: InstructionKind,
    lvalue: AssignmentTarget,
    value: InstrIx,
) -> Result<InstrIx, BuildDiagnostic> {
    Ok(match lvalue {
        AssignmentTarget::Pattern(lvalue) => match lvalue {
            Pattern::Identifier(lvalue) => {
                let identifier = lower_identifier_for_assignment(env, builder, kind, *lvalue)?;
                builder.push(InstructionValue::StoreLocal(hir::StoreLocal {
                    lvalue: LValue { identifier, kind },
                    value: Operand {
                        ix: value,
                        effect: None,
                    },
                }))
            }
            _ => todo!("lower assignment pattern for {:#?}", lvalue),
        },
        _ => todo!("lower assignment for {:#?}", lvalue),
    })
}

fn lower_identifier_for_assignment<'a>(
    _env: &'a Environment<'a>,
    builder: &mut Builder<'a>,
    _kind: InstructionKind,
    identifier: estree::Identifier,
) -> Result<IdentifierOperand<'a>, BuildDiagnostic> {
    let binding = builder.resolve_binding(&identifier)?;
    match binding {
        Binding::Module(..) | Binding::Global => Err(BuildDiagnostic::new(
            DiagnosticError::ReassignedGlobal,
            ErrorSeverity::InvalidReact,
            identifier.range,
        )),
        Binding::Local(id) => Ok(IdentifierOperand {
            identifier: id,
            effect: None,
        }),
    }
}

/// Converts an ESTree literal into a HIR primitive
fn lower_primitive<'a>(
    env: &'a Environment<'a>,
    _builder: &mut Builder<'a>,
    literal: Literal,
) -> PrimitiveValue<'a> {
    match literal.value {
        JsValue::Bool(bool) => PrimitiveValue::Boolean(bool),
        JsValue::Null => PrimitiveValue::Null,
        JsValue::Number(value) => PrimitiveValue::Number(f64::from(value).into()),
        JsValue::String(s) => PrimitiveValue::String(String::from_str_in(&s, &env.allocator)),
        _ => todo!("Lower literal {literal:#?}"),
    }
}
