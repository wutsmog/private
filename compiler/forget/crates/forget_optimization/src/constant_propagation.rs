use std::collections::HashMap;

use forget_diagnostics::Diagnostic;
use forget_estree::{BinaryOperator, JsValue};
use forget_hir::{
    initialize_hir, merge_consecutive_blocks, BlockKind, Environment, Function, GotoKind,
    GotoTerminal, IdentifierId, Instruction, InstructionValue, LoadGlobal, Operand, Primitive,
    TerminalValue,
};
use forget_ssa::eliminate_redundant_phis;

pub fn constant_propagation(env: &Environment, fun: &mut Function) -> Result<(), Diagnostic> {
    let mut constants = Constants::new();
    constant_propagation_impl(env, fun, &mut constants)
}

fn constant_propagation_impl(
    env: &Environment,
    fun: &mut Function,
    constants: &mut Constants,
) -> Result<(), Diagnostic> {
    loop {
        let have_terminals_changed = apply_constant_propagation(env, fun, constants)?;
        if !have_terminals_changed {
            break;
        }
        // If terminals have changed then blocks may have become newly unreachable,
        // so reinitialize the HIR
        // TODO handle errors
        initialize_hir(&mut fun.body).unwrap();

        // Now that predecessors have changed, prune phi operands for unreachable blocks
        // for example, a phi node whose operand was eliminated because it was set in a
        // block that is no longer reached
        for block in fun.body.blocks.iter_mut() {
            // TODO: avoid the clone here
            let predecessors = block.predecessors.clone();
            for phi in block.phis.iter_mut() {
                phi.operands
                    .retain(|predecessor, _| predecessors.contains(predecessor))
            }
        }

        // By removing some phi operands, there may be phis that were not previously
        // redundant but now are
        eliminate_redundant_phis(env, fun);

        // Finally, merge together any blocks that are now guaranteed to execute
        // consecutively
        merge_consecutive_blocks(env, fun)?;
    }
    Ok(())
}

fn apply_constant_propagation(
    env: &Environment,
    fun: &mut Function,
    constants: &mut Constants,
) -> Result<bool, Diagnostic> {
    let mut has_changes = false;

    for block in fun.body.blocks.iter_mut() {
        for phi in block.phis.iter() {
            let mut value: Option<Constant> = None;
            for (_, operand) in &phi.operands {
                if let Some(operand_value) = constants.get(&operand.id) {
                    match &mut value {
                        Some(value) if value == operand_value => {
                            // no-op
                        }
                        Some(_) => {
                            value = None;
                            break;
                        }
                        None => {
                            value = Some(operand_value.clone());
                        }
                    }
                } else {
                    // This phi operand's value is unknown, bail out of replacing it
                    value = None;
                    break;
                }
            }
            if let Some(value) = value {
                constants.insert(phi.identifier.id, value);
            }
        }
        for (ix, instr_ix) in block.instructions.iter().enumerate() {
            if block.kind == BlockKind::Sequence && ix == block.instructions.len() - 1 {
                // Evaluating the last value of a sequence can break order of evaluation
                // so skip these instructions
                continue;
            }
            let instr_ix = usize::from(*instr_ix);
            let mut instr = std::mem::replace(
                &mut fun.body.instructions[instr_ix].value,
                InstructionValue::Tombstone,
            );
            evaluate_instruction(env, &fun.body.instructions, &mut instr, constants)?;
            fun.body.instructions[instr_ix].value = instr;
        }

        // If the block ends in an `if` and the test value is a constant primitive,
        // then convert the terminal into a goto to either the consequent or alternate
        // in this case, only the selected branch is reachable
        if let TerminalValue::If(terminal) = &mut block.terminal.value {
            if let Some(primitive) =
                read_primitive_instruction(&fun.body.instructions, &terminal.test)
            {
                let target_block_id = if primitive.value.is_truthy() {
                    terminal.consequent
                } else {
                    terminal.alternate
                };
                block.terminal.value = TerminalValue::Goto(GotoTerminal {
                    block: target_block_id,
                    kind: GotoKind::Break,
                });
                has_changes = true;
            }
        }
    }

    Ok(has_changes)
}

fn read_primitive_instruction(instrs: &[Instruction], operand: &Operand) -> Option<Primitive> {
    let instr = &instrs[usize::from(operand.ix)].value;
    if let InstructionValue::Primitive(primitive) = instr {
        Some(primitive.clone())
    } else {
        None
    }
}

fn evaluate_instruction(
    env: &Environment,
    instrs: &[Instruction],
    mut instr: &mut InstructionValue,
    constants: &mut Constants,
) -> Result<(), Diagnostic> {
    let read_constant = |operand: &Operand| {
        let instr = &instrs[usize::from(operand.ix)].value;
        match instr {
            InstructionValue::Primitive(value) => Some(Constant::Primitive(value.clone())),
            InstructionValue::LoadGlobal(value) => Some(Constant::Global(value.clone())),
            _ => None,
        }
    };
    match &mut instr {
        InstructionValue::Binary(value) => {
            let left = read_primitive_instruction(instrs, &value.left);
            let right = read_primitive_instruction(instrs, &value.right);
            match (left, right) {
                (Some(left), Some(right)) => {
                    if let Some(result) = apply_binary_operator(env, left, value.operator, right) {
                        *instr = InstructionValue::Primitive(result);
                    }
                }
                _ => {
                    // no-op, not all operands are known
                }
            }
        }
        InstructionValue::LoadLocal(value) => {
            if let Some(const_value) = constants.get(&value.place.identifier.id) {
                *instr = const_value.into();
            }
        }
        InstructionValue::StoreLocal(value) => {
            if let Some(const_value) = read_constant(&value.value) {
                constants.insert(value.lvalue.identifier.identifier.id, const_value);
            }
        }
        InstructionValue::Function(value) => {
            // TODO: due to the outer fixpoint iteration this could visit the same
            // function many times. However we only strictly have to visit the function
            // again if the context variable's constant values have changed since last
            // time.
            // Instead, we can:
            // - Create a filtered Constants instance that extracts just the values for
            //   the function (using its context variables list)
            // - Track the last such filtered Constants instance we visited the function
            //   with. Only visit again if the Constants have changed.
            let mut inner_constants: Constants = value
                .lowered_function
                .context
                .iter()
                .filter_map(|id| {
                    let value = constants.get(&id.identifier.id);
                    value.map(|value| (id.identifier.id, value.clone()))
                })
                .collect();
            constant_propagation_impl(env, &mut value.lowered_function, &mut inner_constants)?;
        }
        _ => {
            // no-op, not all instructions can be processed
        }
    }
    Ok(())
}

fn apply_binary_operator(
    _env: &Environment,
    left: Primitive,
    operator: BinaryOperator,
    right: Primitive,
) -> Option<Primitive> {
    match (left.value, right.value) {
        (JsValue::Number(left), JsValue::Number(right)) => match operator {
            BinaryOperator::Add => Some(Primitive {
                value: JsValue::Number(left + right),
            }),
            BinaryOperator::Subtract => Some(Primitive {
                value: JsValue::Number(left - right),
            }),
            BinaryOperator::Multiply => Some(Primitive {
                value: JsValue::Number(left * right),
            }),
            BinaryOperator::Divide => Some(Primitive {
                value: JsValue::Number(left / right),
            }),
            BinaryOperator::LessThan => Some(Primitive {
                value: JsValue::Boolean(left < right),
            }),
            BinaryOperator::LessThanOrEqual => Some(Primitive {
                value: JsValue::Boolean(left <= right),
            }),
            BinaryOperator::GreaterThan => Some(Primitive {
                value: JsValue::Boolean(left > right),
            }),
            BinaryOperator::GreaterThanOrEqual => Some(Primitive {
                value: JsValue::Boolean(left >= right),
            }),
            BinaryOperator::Equals => Some(Primitive {
                value: JsValue::Boolean(left.equals(right)),
            }),
            BinaryOperator::NotEquals => Some(Primitive {
                value: JsValue::Boolean(left.not_equals(right)),
            }),
            BinaryOperator::StrictEquals => Some(Primitive {
                value: JsValue::Boolean(left.equals(right)),
            }),
            BinaryOperator::NotStrictEquals => Some(Primitive {
                value: JsValue::Boolean(left.not_equals(right)),
            }),
            _ => None,
        },
        (left, right) => match operator {
            BinaryOperator::Equals => left.loosely_equals(&right).map(|value| Primitive {
                value: JsValue::Boolean(value),
            }),
            BinaryOperator::NotEquals => left.not_loosely_equals(&right).map(|value| Primitive {
                value: JsValue::Boolean(value),
            }),
            BinaryOperator::StrictEquals => Some(Primitive {
                value: JsValue::Boolean(left.strictly_equals(&right)),
            }),
            BinaryOperator::NotStrictEquals => Some(Primitive {
                value: JsValue::Boolean(left.not_strictly_equals(&right)),
            }),
            _ => None,
        },
    }
}

type Constants = HashMap<IdentifierId, Constant>;

#[derive(Debug, Clone, PartialEq, Eq)]
enum Constant {
    Global(LoadGlobal),
    Primitive(Primitive),
}

impl From<&Constant> for InstructionValue {
    fn from(value: &Constant) -> Self {
        match value {
            Constant::Global(value) => InstructionValue::LoadGlobal(value.clone()),
            Constant::Primitive(value) => InstructionValue::Primitive(value.clone()),
        }
    }
}

impl From<Constant> for InstructionValue {
    fn from(value: Constant) -> Self {
        match value {
            Constant::Global(value) => InstructionValue::LoadGlobal(value),
            Constant::Primitive(value) => InstructionValue::Primitive(value),
        }
    }
}
