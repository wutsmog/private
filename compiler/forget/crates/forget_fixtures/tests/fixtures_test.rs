use std::env;
use std::fmt::Write;

use bumpalo::Bump;
use forget_build_hir::build;
use forget_estree::{ModuleItem, Statement};
use forget_estree_swc::parse;
use forget_hir::{Environment, Features, Print, Registry};
use forget_optimization::constant_propagation;
use forget_ssa::{eliminate_redundant_phis, enter_ssa};
use insta::{assert_snapshot, glob};
use miette::{NamedSource, Report};

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        println!("ok parse");

        let mut output = String::new();

        for (ix, item) in ast.body.into_iter().enumerate() {
            if let ModuleItem::Statement(stmt) = item {
                if let Statement::FunctionDeclaration(fun) = stmt {
                    let allocator = Bump::new();
                    let environment = allocator.alloc(Environment::new(
                        &allocator,
                        Features {
                            validate_frozen_lambdas: true,
                        },
                        Registry,
                    ));
                    if ix != 0 {
                        output.push_str("\n\n");
                    }
                    match build(&environment, fun.function) {
                        Ok(mut fun) => {
                            // println!("{fun:#?}");
                            // let mut out = String::new();
                            // fun.print(&fun.body, &mut out).unwrap();
                            // println!("{out}");
                            println!("ok build");
                            enter_ssa(&environment, &mut fun).unwrap();
                            println!("ok enter_ssa");
                            eliminate_redundant_phis(&environment, &mut fun);
                            println!("ok eliminate_redundant_phis");
                            constant_propagation(&environment, &mut fun);
                            println!("ok constant_propagation");
                            fun.print(&fun.body, &mut output).unwrap();
                            println!("ok print");
                        }
                        Err(error) => {
                            write!(&mut output, "{}", error,).unwrap();
                            eprintln!(
                                "{:?}",
                                Report::new(error).with_source_code(NamedSource::new(
                                    path.to_string_lossy(),
                                    input.clone(),
                                ))
                            );
                            continue;
                        }
                    };
                }
            }
        }

        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
