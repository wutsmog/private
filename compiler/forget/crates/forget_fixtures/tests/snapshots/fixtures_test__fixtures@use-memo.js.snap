---
source: crates/forget_fixtures/tests/fixtures_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_fixtures/tests/fixtures/use-memo.js
---
Input:
function Component(x) {
  const x = useMemo(() => {
    return y;
  });
  return x;
}


Output:
function Component(
  unknown x$10,
)
entry bb0
bb0 (block)
  [0] unknown $1 = LoadGlobal useMemo
  [1] unknown $14 = DeclareLocal Let unknown t$12
  [2] Label block=bb1 fallthrough=bb6
bb1 (block)
  predecessors: bb0
  [3] unknown $2 = LoadGlobal y
  [4] unknown $13 = StoreLocal Reassign unknown t$12 = unknown $2
  [5] Goto bb6
bb6 (block)
  predecessors: bb1
  [6] unknown $5 = LoadLocal unknown t$12
  [7] unknown $7 = StoreLocal Const unknown x$11 = unknown $5
  [8] unknown $8 = LoadLocal unknown x$11
  [9] Return unknown $8
